frontend/src/pages/projects/screens/detail/connections/ManageConnectionsModal.tsx

```diff
@@ -0,0 +1,154 @@
+import { useK8sNameDescriptionFieldData } from '~/concepts/k8s/K8sNameDescriptionField/K8sNameDescriptionField';
+import { assembleConnectionSecret, getDefaultValues } from '~/concepts/connectionTypes/utils';
+
+type Props = {
+  connection?: Connection;
+  connectionTypes: ConnectionTypeConfigMapObj[];
+  project: ProjectKind;
+  onClose: (submitted?: boolean) => void;
+  onSubmit: (connection: Connection) => Promise<SecretKind>;
+};
+
+export const ManageConnectionModal: React.FC<Props> = ({
+  connection,
+  connectionTypes,
+  project,
+  onClose,
+  onSubmit,
+}) => {
+  const [error, setError] = React.useState<Error>();
+  const [isSaving, setIsSaving] = React.useState(false);
+
+  const enabledConnectionTypes = React.useMemo(
+    () =>
+      connectionTypes.filter((t) => t.metadata.annotations?.['opendatahub.io/enabled'] === 'true'),
+    [connectionTypes],
+  );
+
+  const [selectedConnectionType, setSelectedConnectionType] = React.useState<
+    ConnectionTypeConfigMapObj | undefined
+  >(enabledConnectionTypes.length === 1 ? enabledConnectionTypes[0] : undefined);
+  const { data: nameDescData, onDataChange: setNameDescData } = useK8sNameDescriptionFieldData();
+  const [connectionValues, setConnectionValues] = React.useState<{
+    [key: string]: ConnectionTypeValueType;
+  }>(() => {
+    if (connection?.data) {
+      return connection.data;
+    }
+    if (enabledConnectionTypes.length === 1) {
+      return getDefaultValues(enabledConnectionTypes[0]);
+    }
+    return {};
+  });
+
+  // if user changes connection types, don't discard previous entries in case of accident
+  const previousValues = React.useRef<{
+    [connectionTypeName: string]: {
+      [key: string]: ConnectionTypeValueType;
+    };
+  }>({});
+  const changeSelectionType = React.useCallback(
+    (type?: ConnectionTypeConfigMapObj) => {
+      // save previous connection values
+      if (selectedConnectionType) {
+        previousValues.current[selectedConnectionType.metadata.name] = connectionValues;
+        // clear previous values
+        setConnectionValues({});
+      }
+      // load saved values?
+      if (type?.metadata.name && type.metadata.name in previousValues.current) {
+        setConnectionValues(previousValues.current[type.metadata.name]);
+      } else if (type) {
+        // first time load, so add default values
+        setConnectionValues(getDefaultValues(type));
+      }
+
+      setSelectedConnectionType(type);
+    },
+    [selectedConnectionType, connectionValues],
+  );
+
+  const isValid = React.useMemo(
+    () =>
+      !!selectedConnectionType &&
+      !!nameDescData.name &&
+      !selectedConnectionType.data?.fields?.find(
+        (field) =>
+          isConnectionTypeDataField(field) &&
+          field.required &&
+          !connectionValues[field.envVar] &&
+          field.type !== ConnectionTypeFieldType.Boolean,
+      ),
+    [selectedConnectionType, nameDescData, connectionValues],
+  );
```

Comment:
Individual form fields have an `onValidate` callback. We need to hook into those such that errors are propagated to here.
You can test this with invalid URIs.